src/transformers/models/dac/modeling_dac.py

```diff
@@ -0,0 +1,616 @@
+            projected_latents_i = self.quantizers[i].decode_code(codebook_indices[:, i, :])
+            projected_latents.append(projected_latents_i)
+
+            quantized_representation_i = self.quantizers[i].out_proj(projected_latents_i)
+            quantized_representation = quantized_representation + quantized_representation_i
+        return quantized_representation, torch.cat(projected_latents, dim=1), codebook_indices
+
+    def from_latents(self, latents: torch.Tensor):
+        """Reconstructs the quantized representation from unquantized latents.
+
+        Args:
+            latents (`torch.Tensor` of shape `(batch_size, total_latent_dimension, time_steps)`):
+                Continuous representation of input after projection.
+
+        Returns:
+            quantized_representation (`torch.Tensor`, (batch_size, dimension, time_steps)):
+                Quantized representation of the full-projected space.
+            quantized_latents (`torch.Tensor`, (batch_size, dimension, time_steps)):
+                Quantized representation of the latent space (continuous representation before quantization).
+        """
+        quantized_representation = 0
+        quantized_latents = []
+        codes = []
+        dims = np.cumsum([0] + [q.codebook_dim for q in self.quantizers])
+
+        n_codebooks = np.where(dims <= latents.shape[1])[0].max(axis=0, keepdims=True)[0]
+        for i in range(n_codebooks):
+            j, k = dims[i], dims[i + 1]
+            quantized_latents_i, codes_i = self.quantizers[i].decode_latents(latents[:, j:k, :])
+            quantized_latents.append(quantized_latents_i)
+            codes.append(codes_i)
+
+            quantized_representation_i = self.quantizers[i].out_proj(quantized_latents_i)
+            quantized_representation = quantized_representation + quantized_representation_i
+
+        return quantized_representation, torch.cat(quantized_latents, dim=1), torch.stack(codes, dim=1)
+
+
+class DacDecoder(nn.Module):
+    """DAC Decoder"""
+    def __init__(
+        self,
+        config: DacConfig,
+    ):
+        super().__init__()
+
+        input_channel = config.latent_dim
+        channels = config.decoder_dim
+        rates = config.decoder_rates
+
+        # Add first conv layer
+        layers = [WNConv1d(input_channel, channels, kernel_size=7, padding=3)]
+
+        # Add upsampling + MRF blocks
+        for i, stride in enumerate(rates):
+            input_dim = channels // 2**i
+            output_dim = channels // 2 ** (i + 1)
+            layers += [DacDecoderBlock(input_dim, output_dim, stride)]
+
+        # Add final conv layer
+        layers += [
+            Snake1d(output_dim),
+            WNConv1d(output_dim, 1, kernel_size=7, padding=3),
+            nn.Tanh(),
+        ]
+
+        self.model = nn.ModuleList(layers)
+
+    def forward(self, input):
+        for model in self.model: 
+            input = model(input)
+        return input
```

Comment:
Should return an instance of `DacDecoderOutput`

Reply:
Should we return an instance of `DacDecoderOutput` in `DacModel.decode` ?

Reply:
Yep that's right!